src/recyclerview/components/StickyHeaders.tsx

```diff
@@ -0,0 +1,210 @@
+} from "react";
+import { Animated, NativeScrollEvent } from "react-native";
+import { FlashListProps } from "../..";
+import { CompatAnimatedView } from "./CompatView";
+import { RecyclerViewManager } from "../RecyclerViewManager";
+import { ViewHolder } from "../ViewHolder";
+
+/**
+ * Props for the StickyHeaders component
+ * @template TItem - The type of items in the list
+ */
+export interface StickyHeaderProps<TItem> {
+  /** Array of indices that should have sticky headers */
+  stickyHeaderIndices: number[];
+  /** The data array being rendered */
+  data: readonly TItem[];
+  /** Animated value tracking scroll position */
+  scrollY: Animated.Value;
+  /** Function to render each item */
+  renderItem: FlashListProps<TItem>["renderItem"];
+  /** Ref to access sticky header methods */
+  stickyHeaderRef: React.RefObject<StickyHeaderRef>;
+  /** Manager for recycler view operations */
+  recyclerViewManager: RecyclerViewManager<TItem>;
+  /** Additional data to trigger re-renders */
+  extraData: FlashListProps<TItem>["extraData"];
+}
+
+/**
+ * Ref interface for StickyHeaders component
+ */
+export interface StickyHeaderRef {
+  /** Reports scroll events to update sticky header positions */
+  reportScrollEvent: (event: NativeScrollEvent) => void;
+}
+
+export const StickyHeaders = <TItem,>({
+  stickyHeaderIndices,
+  renderItem,
+  stickyHeaderRef,
+  recyclerViewManager,
+  scrollY,
+  data,
+  extraData,
+}: StickyHeaderProps<TItem>) => {
+  const [stickyIndices, setStickyIndices] = useState<{
+    currentStickyIndex: number;
+    nextStickyIndex: number;
+  }>({ currentStickyIndex: -1, nextStickyIndex: -1 });
+
+  const { currentStickyIndex, nextStickyIndex } = stickyIndices;
+  const hasLayout = recyclerViewManager.hasLayout();
+
+  // Memoize sorted indices based on their Y positions
+  const sortedIndices = useMemo(() => {
+    if (!hasLayout) {
+      return [];
+    }
+    return stickyHeaderIndices.sort((a, b) => a - b);
+  }, [stickyHeaderIndices, recyclerViewManager, hasLayout]);
+
+  const compute = useCallback(() => {
+    const adjustedValue = recyclerViewManager.getLastScrollOffset();
+
+    // Binary search for current sticky index
+    const currentIndexInArray = findCurrentStickyIndex(
+      sortedIndices,
+      adjustedValue,
+      (index) => recyclerViewManager.getLayout(index).y
+    );
+
+    const newStickyIndex = sortedIndices[currentIndexInArray] ?? -1;
+    let newNextStickyIndex = sortedIndices[currentIndexInArray + 1] ?? -1;
+
+    if (newNextStickyIndex > recyclerViewManager.getEngagedIndices().endIndex) {
+      newNextStickyIndex = -1;
+    }
+
+    if (
+      newStickyIndex !== currentStickyIndex ||
+      newNextStickyIndex !== nextStickyIndex
+    ) {
+      setStickyIndices({
+        currentStickyIndex: newStickyIndex,
+        nextStickyIndex: newNextStickyIndex,
+      });
+    }
+  }, [currentStickyIndex, nextStickyIndex, recyclerViewManager, sortedIndices]);
+
+  useEffect(() => {
```

Comment:
@naqvitalha Maybe the bug we saw is because the current index is computed in a useEffect instead of a layout effect. This would be computed well after pain right?